app/core/versioning.py

```
@@ -0,0 +1,296 @@
+        version = version or self._current_version
+        if not version:
+            raise VersionError("No version specified and no current version set")
+
+        if version not in self._schemas:
+            raise VersionError(f"Unknown version: {version}")
+
+        try:
+            self._validate_against_schema(data, self._schemas[version])
+            return True
+        except SchemaError as e:
+            logger.error(f"Validation error for version {version}: {str(e)}")
```

Ruff (0.8.2)

Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)